File: attractor/pipeline/conditions.py
"""Condition expression language evaluator.

Supports: = (equals), != (not equals), && (AND conjunction).
Variables: outcome, preferred_label, context.* (missing = empty string).
Empty condition always returns True.
"""

from __future__ import annotations

import re
from typing import Any

from .context import PipelineContext
from ..exceptions import AttractorValidationError


def evaluate(
    expression: str,
    context: PipelineContext,
    outcome: str = "",
    preferred_label: str = "",
) -> bool:
    """Evaluate a condition expression against context and outcome."""
    expr = expression.strip()
    if not expr:
        return True

    # Split on && (AND conjunction)
    clauses = [c.strip() for c in expr.split("&&")]
    return all(_eval_clause(c, context, outcome, preferred_label) for c in clauses)


def _eval_clause(
    clause: str,
    context: PipelineContext,
    outcome: str,
    preferred_label: str,
) -> bool:
    """Evaluate a single clause (comparison)."""
    # Try != first (before =)
    if "!=" in clause:
        parts = clause.split("!=", 1)
        if len(parts) == 2:
            left = _resolve_value(parts[0].strip(), context, outcome, preferred_label)
            right = _resolve_value(parts[1].strip(), context, outcome, preferred_label)
            return str(left) != str(right)

    if "=" in clause:
        parts = clause.split("=", 1)
        if len(parts) == 2:
            left = _resolve_value(parts[0].strip(), context, outcome, preferred_label)
            right = _resolve_value(parts[1].strip(), context, outcome, preferred_label)
            return str(left) == str(right)

    raise AttractorValidationError(f"Invalid condition clause: {clause}")


def _resolve_value(
    token: str,
    context: PipelineContext,
    outcome: str,
    preferred_label: str,
) -> str:
    """Resolve a value token to its string value."""
    token = token.strip()

    # Strip quotes if present
    if (token.startswith('"') and token.endswith('"')) or \
       (token.startswith("'") and token.endswith("'")):
        return token[1:-1]

    # Built-in variables
    if token == "outcome":
        return outcome
    if token == "preferred_label":
        return preferred_label

    # Context lookup
    if token.startswith("context."):
        key = token[8:]  # Remove "context."
        val = context.get(key, "")
        return str(val) if val is not None else ""

    # Literal values
    return token


def validate_condition(expression: str) -> list[str]:
    """Validate condition syntax. Returns list of error messages."""
    errors: list[str] = []
    expr = expression.strip()
    if not expr:
        return errors

    clauses = [c.strip() for c in expr.split("&&")]
    for clause in clauses:
        if "!=" in clause:
            parts = clause.split("!=", 1)
        elif "=" in clause:
            parts = clause.split("=", 1)
        else:
            errors.append(f"Clause missing operator (= or !=): '{clause}'")
            continue

        if len(parts) != 2 or not parts[0].strip() or not parts[1].strip():
            errors.append(f"Invalid clause: '{clause}'")

    return errors

File: internal/pipeline/conditions.go
package pipeline

import (
	"fmt"
	"strings"

	"github.com/nigelpepper/attractor/internal/aerr"
)

// EvaluateCondition evaluates a condition expression against context and
// outcome. Supports "=", "!=", and "&&". An empty expression is always true.
func EvaluateCondition(expression string, ctx *PipelineContext, outcome, preferredLabel string) (bool, error) {
	expr := strings.TrimSpace(expression)
	if expr == "" {
		return true, nil
	}
	for _, clause := range strings.Split(expr, "&&") {
		ok, err := evalClause(strings.TrimSpace(clause), ctx, outcome, preferredLabel)
		if err != nil {
			return false, err
		}
		if !ok {
			return false, nil
		}
	}
	return true, nil
}

func evalClause(clause string, ctx *PipelineContext, outcome, preferredLabel string) (bool, error) {
	if strings.Contains(clause, "!=") {
		parts := strings.SplitN(clause, "!=", 2)
		if len(parts) == 2 {
			l := resolveValue(strings.TrimSpace(parts[0]), ctx, outcome, preferredLabel)
			r := resolveValue(strings.TrimSpace(parts[1]), ctx, outcome, preferredLabel)
			return l != r, nil
		}
	}
	if strings.Contains(clause, "=") {
		parts := strings.SplitN(clause, "=", 2)
		if len(parts) == 2 {
			l := resolveValue(strings.TrimSpace(parts[0]), ctx, outcome, preferredLabel)
			r := resolveValue(strings.TrimSpace(parts[1]), ctx, outcome, preferredLabel)
			return l == r, nil
		}
	}
	return false, &aerr.ValidationError{Msg: fmt.Sprintf("Invalid condition clause: %s", clause)}
}

func resolveValue(token string, ctx *PipelineContext, outcome, preferredLabel string) string {
	token = strings.TrimSpace(token)
	if len(token) >= 2 {
		if (strings.HasPrefix(token, `"`) && strings.HasSuffix(token, `"`)) ||
			(strings.HasPrefix(token, "'") && strings.HasSuffix(token, "'")) {
			return token[1 : len(token)-1]
		}
	}
	switch token {
	case "outcome":
		return outcome
	case "preferred_label":
		return preferredLabel
	}
	if strings.HasPrefix(token, "context.") {
		return ctx.GetString(token[len("context."):])
	}
	return token
}

// ValidateCondition checks condition syntax, returning a list of error messages.
func ValidateCondition(expression string) []string {
	var errs []string
	expr := strings.TrimSpace(expression)
	if expr == "" {
		return errs
	}
	for _, clause := range strings.Split(expr, "&&") {
		clause = strings.TrimSpace(clause)
		var parts []string
		switch {
		case strings.Contains(clause, "!="):
			parts = strings.SplitN(clause, "!=", 2)
		case strings.Contains(clause, "="):
			parts = strings.SplitN(clause, "=", 2)
		default:
			errs = append(errs, fmt.Sprintf("Clause missing operator (= or !=): '%s'", clause))
			continue
		}
		if len(parts) != 2 || strings.TrimSpace(parts[0]) == "" || strings.TrimSpace(parts[1]) == "" {
			errs = append(errs, fmt.Sprintf("Invalid clause: '%s'", clause))
		}
	}
	return errs
}
